### DISCUSSION TITLE
Vanishing messages, removed queues

### DISCUSSION THREAD

**Discussion by yt87** [MetPX/sarracenia#813] [Q&A] [answered]
This seems to be a server issue, I am not sure where to report it.
I tried to retrieve grib files for the 10 km regional model. My configuration file is
```
# Fichiers GRIB pour le SRPD 10km
# Pour plus d'information sur ce produit, veuillez consulter la documentation a
#  http://weather.gc.ca/grib/grib2_reg_15km_f.html
#
download True
logMetrics True
logReject True
acceptUnmatched False

broker amqps://dd.weather.gc.ca/
topicPrefix v02.post

# instances:
#    quantité de processus qui téléchargent simultanément. 
instances 4

# expire, en usage opérationnel, c´est mieux 1d (1 jour) pour celui-ci. Il faut que ça soit
# plus long que l´interruption le plus long qu´on veut toléré sans perte.
expire 6h

subtopic model_gem_regional.10km.grib2.#
# FIXME: change this
directory /tmp/metpx/reg
reject .*/[01][0-9]/0([036][^0369]|[147][^258]|[258][^147])/
reject .*_[A-Z]{3,4}_ISBL_(50|[3-5]50|[1-2][27]5)_
accept .*_(TMP|SPFH|DEPR|UGRD|VGRD)_(ISBL|TGL)_
accept .*_(ABSV|VVEL|HGT)_ISBL_
accept .*_CWAT_EATM_
accept .*_PRMSL_
accept .*(4LFTX|SHOWA)_SFC
accept .*(CAPE|HLCY)_ETAL
accept .*_(PRES|ICEC|SNOD|WTMP|AC?PCP|SPFH|TCDC|PRATE|WEA..)_SFC# Fichiers GRIB pour le SRPD 10km
# Pour plus d'information sur ce produit, veuillez consulter la documentation a
#  http://weather.gc.ca/grib/grib2_reg_15km_f.html
#
download True
logMetrics True
logReject True
acceptUnmatched False

broker amqps://dd.weather.gc.ca/
topicPrefix v02.post

# instances:
#    quantité de processus qui téléchargent simultanément. 
instances 4

# expire, en usage opérationnel, c´est mieux 1d (1 jour) pour celui-ci. Il faut que ça soit
# plus long que l´interruption le plus long qu´on veut toléré sans perte.
expire 6h

subtopic model_gem_regional.10km.grib2.#
# FIXME: change this
directory /tmp/metpx/reg
reject .*/[01][0-9]/0([036][^0369]|[147][^258]|[258][^147])/
reject .*_[A-Z]{3,4}_ISBL_(50|[3-5]50|[1-2][27]5)_
accept .*_(TMP|SPFH|DEPR|UGRD|VGRD)_(ISBL|TGL)_
accept .*_(ABSV|VVEL|HGT)_ISBL_
accept .*_CWAT_EATM_
accept .*_PRMSL_
accept .*(4LFTX|SHOWA)_SFC
accept .*(CAPE|HLCY)_ETAL
accept .*_(PRES|ICEC|SNOD|WTMP|AC?PCP|SPFH|TCDC|PRATE|WEA..)_SFC
```
First I ran sr3 on my home desktop, metpx version 3.00.45 with the ``download`` flag set to ``False``. All messages were retrieved (18Z model run). For the same model run, I ran the same configuration on our operational system at work, this time ``download`` set to ``True``. Only a subset of the data files was downloaded. Distribution by forecast hour (period, # retrieved files) :
```
P000 107
P003 12
P006 13
P009 0
P012 126
P015 8
P018 51
P021 76
P024 23
P027 22
P030 0
P033 18
P036 23
P039 8
P042 28
P045 24
P48 30
P51 12
P54 53
P57 12
P60 193
P63 193
P66 193
P69 193
P72 193
P75 193
P78 193
P81 188
P84 193
```
I verified that the received notifications matched the downloaded files. There were no errors, only a few warnings in the logs.

The next attempt was retrieval of the 00Z model run on my home computer. During that time, I monitored the network traffic with ``iftop``. It started well, but after about half an hour the traffic slowed to a crawl and at 4Z it stopped completely. I would guess that the queue on the server was scrubbed. The final count of downloaded files:
```
P000 188
P003 193
P006 193
P009 193
P012 193
P015 193
P018 193
P021 82
P024 142
P027 48
```
Again, there were no errors or even warnings in the log files.

Are there restrictions on the server side for the amount of data a client can download? Or something else is going on?

**Comment by petersilva** [comment]
stating the obvious: no, that's not normal, you should have gotten the full complement every time (presumably 193 on every line.) there are restrictions related to having to high a backlog (10,000 files in the queue or something like that.) but I don't think that would be the problem here.  

note: the configuration seems to be posted twice

**Reply by petersilva** [reply]
I copy/pasted to a home system, wll try to reproduce.

**Comment by yt87** [comment]
I will try again for the 18Z model issue on my home system (faster internet, bypassing NOAA gateway).
Several years ago (~2016) I modified metpx to conform to my own design (logging location, use of ``pycurl`` to pull data...).  A few years later, I ran into the same issue, the queues were removed on the server before transfer completion. Then I switched to cron scheduled jobs retrieving data from https://dd.weather.gc.ca. It worked better, but recently I started getting complaints from our weather offices about missing data. That's why I started experimenting with the ''official'' metpx software.
Also, once I did run in to a strange issue: the retrieval worked for one client, but was failing on another system. I was told (by Anne-Marie Baubien, if I recall) that the algorithm that determines the backend server host is deterministic, based on the client ip. It happened then that the assigned host for my failing client was down. I will try to note down the hosts metpx connects to (there were 3 yesterday).
In case you have access to the server logs: my ip address is 72.42.159.7

**Reply by reidsunderland** [reply]
Queues on our brokers also expire, in addition to the 10,000 message limit that @petersilva mentioned. The config posted above has `expire 6h` , so the queue may be expiring (deleting itself). 

Is your queue name `q_anonymous_subscribe.gem_reg.39122536.41612943`? `sr3 show subscribe/config_name` will show it if you need to check. I can see that that queue is bound to `model_gem_regional.10km.grib2.#` and it's currently around around 10k messages - it seems to be receiving messages faster than it can acknowledge (download). This is likely causing messages to be discarded. 

When you had `download False` set, your subscriber was likely keeping up with the incoming message rate, but maybe now that it's downloading files, it isn't working fast enough. Sarracenia will ack messages once it has downloaded the file. You already have `instances 4` to download in parallel, but you could try increasing that to more instances if you don't think there's a bandwidth problem.


---

You could also try using our other broker to see if you get better results, [hpfx.collab.science.gc.ca](https://hpfx.collab.science.gc.ca/). The same data is available, but the structure is a bit different so you'd have to change your subtopic:

```
broker amqps://anonymous:anonymous@hpfx.collab.science.gc.ca/
# first *. is to subscribe to any date
subtopic *.WXO-DD.model_gem_regional.10km.grib2.#
```

**Comment by yt87** [comment]
At 21:10Z I have received ~1500 files. I see connections to 3 hosts: 105.189.10.90, 91 and 47. Only two are active, the .47 traffic rate is ~30 kb/s all the time. The total volume retrieved is ~0.95 GB. So it is much better than yesterday, when it timed out at about 0.8 GB.
I thought ``expire 6h`` should be enough for model cycle being 6 hour. I'll try hpfx.collab for the next run.

**Reply by reidsunderland** [reply]
The .90 and .91 addresses are the HTTP servers where the data is coming from, and .47 is the AMQP broker, which only sends the very small messages.

Expire 6h may be enough and it could just be the 10,000 message limit causing some messages to be missed.

**Comment by yt87** [comment]
Forgot about the queue. Yest, it is ``'resolved_qname': 'q_anonymous_subscribe.gem_reg.39122536.41612943'``. 3080 files received at 21:32Z, but connections to .90 and .91 are dropped, no data coming in. I started sr3 about 2.5 hours ago.

**Comment by reidsunderland** [comment]
@petersilva , would [`prefetch` ](https://metpx.github.io/sarracenia/Reference/sr3_options.7.html#prefetch-n-default-1) help if the queue is reaching 10,000 messages on the broker and downloads are going too slow?

**Reply by petersilva** [reply]
I don't think prefetch will make much difference, it defaults to 25, could raise it if you want,... I dunno 200?

**Comment by yt87** [comment]
Final result same format as before:
```
P000 175
P003 74
P006 42
P009 49
P012 193
P015 10
P018 46
P021 62
P024 44
P027 61
P030 6
P033 7
P036 58
P039 2
P042 127
P045 78
P048 49
P051 47
P054 42
P057 114
P060 193
P063 193
P066 193
P069 193
P072 193
P075 193
P078 193
P081 193
P084 193
```
Interestingly, the earlier files are missing. Note that the total number of files for this model is about 32,000, all come within ~15 min. The 10k limit for number of messages is to small, unless the client is on a fast network.

**Comment by petersilva** [comment]
Can someone check the cron jobs on dd.weather? I have a feeling that if one is more than 3 hours behind, the files get deleted at source. then attempts to download, fail, and it will get queued for retry for days, and that will slow things down.  perhaps do *sr3 status* on your subscribers when downloading.  There should be a retry queue listed. see what it looks like, also perhaps consult the *lag* display.

**Reply by petersilva** [reply]
ok ran it at home here are results for yesterday:
```

fractal% ./summary.py 20231107
 malformed filename skipped: kk
examined
 tally for 2023110712
P078 : 182
P063 : 182
P057 : 182
P009 : 182
P024 : 182
P039 : 182
P015 : 182
P045 : 182
P018 : 182
P036 : 182
P066 : 182
P012 : 182
P000 : 182
P042 : 182
P027 : 182
P069 : 182
P060 : 182
P003 : 182
P051 : 182
P081 : 182
P048 : 182
P033 : 182
P054 : 182
P072 : 182
P021 : 182
P030 : 182
P075 : 182
P006 : 182
P084 : 182
 tally for 2023110718
P063 : 182
P060 : 182
P048 : 182
P012 : 182
P030 : 182
P027 : 182
P078 : 182
P075 : 182
P021 : 182
P018 : 182
P072 : 182
P009 : 182
P015 : 182
P033 : 182
P045 : 182
P051 : 182
P000 : 182
P042 : 182
P054 : 182
P057 : 182
P084 : 182
P069 : 182
P003 : 182
P036 : 182
P039 : 182
P066 : 182
P006 : 182
P024 : 182
P081 : 182
fractal% 

```
today (so far):

```
ractal% ./summary.py 20231108
examined
 tally for 2023110800
P084 : 182
P054 : 182
P063 : 182
P081 : 182
P048 : 182
P030 : 182
P018 : 182
P072 : 182
P033 : 182
P057 : 182
P060 : 182
P021 : 182
P066 : 182
P036 : 182
P051 : 182
P000 : 182
P039 : 182
P045 : 182
P075 : 182
P006 : 182
P069 : 182
P015 : 182
P003 : 182
P024 : 182
P042 : 182
P009 : 182
P078 : 182
P027 : 182
P012 : 182
fractal% 

```

so far, It looks fine.  I'm doing this on a home PC, on a home ISP, use four instances also.
It seems to completely work for me.

**Reply by petersilva** [reply]
fwiw... code to do the summary based on the download directories:

```python
#!/usr/bin/python3

import os
import sys


tally={}
for i in sys.argv[1:]:

    for grib_output in os.listdir(i):

        # CMC_reg_UGRD_ISBL_10_ps10km_2023110800_P030.grib2
        #  0   1   2    3    4   5     6          7
        fields= grib_output.replace('.grib2','').split('_')

        if len(fields) < 7:
            print ( f" malformed filename skipped: {grib_output}" )
            continue

        run = fields[6]
        t=fields[7]

        if run not in tally:
            tally[run]={}

        if t not in tally[run]:
            tally[run][t] = 1
        else:
            tally[run][t] += 1


print('examined')

for r in tally:
    print( f" tally for {r}" )
    for t in tally[r]:
       print( f"{t} : {tally[r][t]}" )
```

**Comment by petersilva** [comment]
To tell the truth, I never liked having the queue limit at all.  We were forced to do something because we had people starting queues and leaving them there, where they would build up to millions, and slow the service for everyone... kind of a inadvertant DOS attacks.    The analysts on duty picked 10,000 and we thought about it a bit, bit it was really just a guess on our part.  It could very well be that a higher number is needed.

**Comment by petersilva** [comment]
can someone clarify why this is closed? I'm not sure if it's resolved... whoever closed it, please mark their reasoning.  I'm worried I closed it myself with a mis-click.

**Reply by yt87** [reply]
Not me.

**Reply by petersilva** [reply]
ok, must have been me... I re-opened it.

**Reply by petersilva** [reply]
on issues, it says who did what, but discussions don't ... noted.

**Reply by petersilva** [reply]
fwiw: 6Z run is clean for me as well.

**Comment by petersilva** [comment]
12Z runs: they all have 183... 1 extra?   at any rate the results seem fairly consistent.
... so it does not look like it is a problem if the subscriber can keep up.  If you have the chance, you should run:

```

sr3 status

```
While the downloads are happenning, and look at how much lag is reported. the more lag, the more likely stuff will get dropped.

**Comment by yt87** [comment]
Lags were huge, 1000+ s.  I am pretty sure it is the queue limit issue.  What I don't understand is why there were files missing from earlier forecast hours, the later periods came OK. It looks like the queue was purged when the limit was hit.  
I gave up on trying. In the short term, we might get data from the Eastern Region. I don't know how it works for them. I am trying to understand the code, maybe I can come up with something - a longer term goal.

**Reply by reidsunderland** [reply, marked as answer]
I believe when the queue limit is reached, the oldest messages are purged from the queue and the 10,000 most recently published will remain. 

You could further tweak the subtopics a bit to reduce the number of messages that get put into your queue. Since you only want every third hour, you could try

```
subtopic model_gem_regional.10km.grib2.*.000.#
subtopic model_gem_regional.10km.grib2.*.003.#
subtopic model_gem_regional.10km.grib2.*.006.#
subtopic model_gem_regional.10km.grib2.*.009.#
subtopic model_gem_regional.10km.grib2.*.012.#
subtopic model_gem_regional.10km.grib2.*.015.#
subtopic model_gem_regional.10km.grib2.*.018.#
subtopic model_gem_regional.10km.grib2.*.021.#
subtopic model_gem_regional.10km.grib2.*.024.#
subtopic model_gem_regional.10km.grib2.*.027.#
subtopic model_gem_regional.10km.grib2.*.030.#
subtopic model_gem_regional.10km.grib2.*.033.#
subtopic model_gem_regional.10km.grib2.*.036.#
subtopic model_gem_regional.10km.grib2.*.039.#
subtopic model_gem_regional.10km.grib2.*.042.#
subtopic model_gem_regional.10km.grib2.*.045.#
subtopic model_gem_regional.10km.grib2.*.048.#
subtopic model_gem_regional.10km.grib2.*.051.#
subtopic model_gem_regional.10km.grib2.*.054.#
subtopic model_gem_regional.10km.grib2.*.057.#
subtopic model_gem_regional.10km.grib2.*.060.#
subtopic model_gem_regional.10km.grib2.*.063.#
subtopic model_gem_regional.10km.grib2.*.066.#
subtopic model_gem_regional.10km.grib2.*.069.#
subtopic model_gem_regional.10km.grib2.*.072.#
subtopic model_gem_regional.10km.grib2.*.075.#
subtopic model_gem_regional.10km.grib2.*.078.#
subtopic model_gem_regional.10km.grib2.*.081.#
subtopic model_gem_regional.10km.grib2.*.084.#
```

If you want to try this, do `sr3 stop subscribe/config_name` and `sr3 cleanup subscribe/config_name` to ensure your old queue (with its old subtopic) gets deleted from our broker.

A bit of a hack that may also help is to have multiple configs and divide the subtopics between each one. If you're expecting to get 32,000 files, maybe split up the subtopics above into 4 different configs. Each config will have its own queue, and should receive 32,000/4 = 8000 messages.

**Reply by yt87** [reply]
Thanks, this works for me. My effective transfer rate is about 1MB/s with 4 instances. The total volume of subset of files we are interested in is ~100MB per each forecast period. You can do the math. We also want gem global, for all hours.
So, multiple queues are the solution for now. I think we can close this discussion until a less hackish solution can be found. I'll keep looking.

**Reply by petersilva** [reply]
@reidsunderland, @junhu3  I think we should review and confirm:
* is the current limit 10,000 ... can someone confirm the policy settings?
* Should we raise the limit from 10,000 to 50,000 ?

rule of thumb:

We should assume that for every client patient enough to work through the problem, there are 9 others who assume it's broken and give up.

We need to pick thresholds that allow for reasonable use cases while providing protection from abuse/misuse.

**Reply by reidsunderland** [reply]
Yes, the limit is currently 10,000.

**Comment by petersilva** [comment]
I'm going to ask about raising the limit to 50,000

**Comment by petersilva** [comment]
The other thing is, maybe you ( @yt87 )need to talk to your networking people to inquire about bandwidth between us.  1MB is kinda slow.  Since getting it to my house is no problem... (meaning it exits all Canadian infrastructure) the bottleneck needs to be further away from us.
might want to start some traceroutes, and such.

**Reply by yt87** [reply]
A few years ago, we were receiving polar orbiter satellite data from the University of Alaska in Fairbanks. Our Fairbanks WFO is located in the same building as the University processing centre. Due to security policies, the feed had to be routed through NWS gateway in the suburbs of DC. The time lag was hours, a lot of data were dropped. It took me several weeks to diagnose and find a workaround to the dropped products: increase some parameters in LDM configuration. Currently we are receiving raw data  (a smaller volume) and do processing locally, so the issue went away.  The network bandwidth remains the same.
 I have faster connection at home than through the public gateway at work. It is still not fast enough for this particular route. I can download other data at much faster speed, so my suspicion is that the bottleneck is on the border. Antiquated equipment at CISA ow NSA maybe?
With regard to GEM files, it is another absurd. All five or six (not sure about Pacific) regions download the files on a public network. All started having download issues recently (I doubt metpx is used by anyone except Alaska). NCEP has a direct fast connection to CMC. It would make sense for them to pull the data and distribute to all forecast offices. But no, it can't be done.
I left AES because of frustration with the way things are done in public service. Certainly it is not better here.

**Reply by petersilva** [reply]
for what it's worth:

"NCEP has a direct fast connection to CMC" ... no. there is no such thing.  We have no direct connections to any NWS/NOAA/U.S. government sites. There are no dedicated lines any more. Everything goes through government-wide centralized points of presence. ECCC doesn't have it's own internet for the last oh... again... around 10 years.  We have two links that eventually get to the internet:  normal public internet, and Internet2/Canarie. The traffic between NWS and NCEP could be taking one or the other.  We can't know without going to the end point in question, and doing a traceroute.

When you say that multiple centres are downloading... they are all using the same internet links on the Canadian side.  I don't know how NWS is setup, but from your Alaska story, it sounds like it's the same with them: a centralized single Point of Presence for internet access with heavy security instrumentation.  Ideally, the model data would traverse that PoP once. and someone in NWS would re-distribute. I think we agree on that.  That's how we ingest external data for our storm prediction centres.

so... you mentioned you were happy enough with @reidsunderland's topics that you had what you needed... we can close the discussion here for now...you can email ECCC directly (see email later in this thread.) to get better/faster access, and we will be improving our settings so fewer people have these problems in future.  

Thanks very much for asking your questions, and please ask again when it makes sense to do so.

**Comment by petersilva** [comment]
@yt87 if you send an email to ec.dps-client.ec@canada.ca then we can set you up with an account besides anonymous...
and it won't have the 10,000 limit.

**Comment by petersilva** [comment]
after internal discussions, there was agreement in principle to raise the limit to 50,000.  Such a change has to go through change management, will probably take a couple of weeks before it happens.

**Comment by petersilva** [comment]
Isn't it weird that if you close a discussion, it doesn't show up in the Discussions list? 
I'm confused... isn't the point of this so that other people can find it?
hmm...
anyways closing...

**Reply by petersilva** [reply]
I can't figure out whether closing a discussion when it comes to a natural end is what we should do, or just leave it open... I felt odd closing it, so re-opened...

**Reply by reidsunderland** [reply]
I changed the discussion type to Q&A and marked my own response as the answer. Now it has a nice green checkmark and says answered in the list, but it's still open..

**Reply by yt87** [reply]
I closed #811 for a test: it does show in the list, but you must change filter to All or Closed. The default is Open.

**Reply by petersilva** [reply]
I like the green checkmark... I guess the default is to encourage participation in active discussions... and keep them separate from ones that are done.  I guess it's the same for issues (default search only looks in open issues also...) It still feels weird to close it...

**Comment by petersilva** [comment]
Changes that should make things easier in future:

* With the year old sr3, there was a pathological behaviour on retries:  If a download failed, it would queue and retry again after a while... and keep doing that for 3 days.  Current version (v3.0.56) looks at the upstream server, and if the file isn't there, it just says... It's gone, and does not try to download again.  The most common case of "retries" was the file being deleted upstream, and then retries accumulating.  The effort of doing the retries also slowed down the current downloads.  Not queuing for retry accellerates normal downloads.
* Today, there is a new version of the dd tree.   With the old scheme some data sets were gone after as little as three hours. Now upstream data should ALWAYS be present for several days.  The phenomenon of falling a few hours behind, and then not being able to get the data at all should be gone  now.  You will always have several days to download.
